css/css-view-transitions/pseudo-element-animations fails to read 'opacity' computed style.

mattwoodrow opened the pull request:
#### ab265bfb1c613dad4684d189f2ae2d977d57e213
<pre>
css/css-view-transitions/pseudo-element-animations fails to read &apos;opacity&apos; computed style.
<a href="https://bugs.webkit.org/show_bug.cgi?id=293031">https://bugs.webkit.org/show_bug.cgi?id=293031</a>
&lt;<a href="https://rdar.apple.com/151351674">rdar://151351674</a>&gt;

Reviewed by Antoine Quint.

The test is getting the computed style, and trying to read opacity, which should
work since that&apos;s being animated.

This is because computeRenderStyleForProperty detects that &apos;opacity&apos; could be
accelerated, and tries to use the originating element&apos;s animated style (and then
the cached pseudo style off of it).

Change this code to actually lookup the keyframes and check for an accelerated
animation of the current property, and then access the animated style (of the
pseudo&apos;s renderer, not the originating element&apos;s) if so.

I am unsure why this test now dumps the test file into the output, but we pass
an extra check, and I&apos;ll followup to fix the remaining failure (which then makes
the weird output go away).

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-element-animations-expected.txt:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::computeRenderStyleForProperty):
(WebCore::ComputedStyleExtractor::customPropertyValue const):
(WebCore::ComputedStyleExtractor::propertyValue const):
(WebCore::ComputedStyleExtractor::layerCount const):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::isRunningAcceleratedTransformAnimation const):
(WebCore::Styleable::isRunningAcceleratedAnimationOfProperty const):
* Source/WebCore/style/Styleable.h:

Canonical link: <a href="https://commits.webkit.org/295141@main">https://commits.webkit.org/295141@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d6d6631c3a6379860b5d04b29c8451e06824d5a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104205 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23909 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14236 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109404 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54873 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106245 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24277 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32453 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/79145 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107211 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18861 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94010 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59472 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12047 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54233 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12106 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111787 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31361 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/23120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/88166 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31725 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90198 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87827 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22358 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32720 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/10469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25825 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31290 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36603 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/31084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34420 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/32644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/style/StyleExtractor.cpp:
```diff
@@ -199,18 +199,14 @@ bool Extractor::updateStyleIfNeededForProperty(Element& element, CSSPropertyID p
-            return ownedStyle->getCachedPseudoStyle(*pseudoElementIdentifier);
+    if (Style::Interpolation::isAccelerated(propertyID, element.document().settings())) {
+        Styleable styleable(element, pseudoElementIdentifier);
+        if (styleable.isRunningAcceleratedAnimationOfProperty(propertyID) && styleable.renderer()) {
```
I'd move the second check (`styleable.renderer()`) to be performed first since it's bound to be cheaper.

graouts commented on Source/WebCore/style/Styleable.h:
```diff
@@ -80,6 +80,7 @@ struct Styleable {
     bool mayHaveNonZeroOpacity() const;
 
     bool isRunningAcceleratedTransformAnimation() const;
+    bool isRunningAcceleratedAnimationOfProperty(CSSPropertyID) const;
```
Perhaps we should get rid of `isRunningAcceleratedTransformAnimation()` and replace it with a call to `isRunningAcceleratedAnimationOfProperty(CSSPropertyTransform)`? I have no idea if there are many call sites. Fine to do as a followup of course.

mattwoodrow committed:
css/css-view-transitions/pseudo-element-animations fails to read 'opacity' computed style.
https://bugs.webkit.org/show_bug.cgi?id=293031
<rdar://151351674>

Reviewed by Antoine Quint.

The test is getting the computed style, and trying to read opacity, which should
work since that's being animated.

This is because computeRenderStyleForProperty detects that 'opacity' could be
accelerated, and tries to use the originating element's animated style (and then
the cached pseudo style off of it).

Change this code to actually lookup the keyframes and check for an accelerated
animation of the current property, and then access the animated style (of the
pseudo's renderer, not the originating element's) if so.

I am unsure why this test now dumps the test file into the output, but we pass
an extra check, and I'll followup to fix the remaining failure (which then makes
the weird output go away).

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-element-animations-expected.txt:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::computeRenderStyleForProperty):
(WebCore::ComputedStyleExtractor::customPropertyValue const):
(WebCore::ComputedStyleExtractor::propertyValue const):
(WebCore::ComputedStyleExtractor::layerCount const):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::isRunningAcceleratedTransformAnimation const):
(WebCore::Styleable::isRunningAcceleratedAnimationOfProperty const):
* Source/WebCore/style/Styleable.h:

Canonical link: https://commits.webkit.org/295141@main
